[SPARK-2883][SQL] Orc support through datasource api

scwf opened the pull request:
Adding support for read/write orc files through the new datasource api.

scwf committed:
Orc support through datasource api

scwf committed:
style fix

SparkQA commented:
[Test build #27469 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27469/consoleFull) for   PR 3753 at commit [`9d7c082`](https://github.com/apache/spark/commit/9d7c0826a5d021015b4382559919ac1e1d582d31).
 * This patch merges cleanly.

SparkQA commented:
[Test build #27470 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27470/consoleFull) for   PR 3753 at commit [`f21b693`](https://github.com/apache/spark/commit/f21b6936e84c98cea69c2ed7835313fcb01dcf71).
 * This patch merges cleanly.

SparkQA commented:
[Test build #27470 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27470/consoleFull) for   PR 3753 at commit [`f21b693`](https://github.com/apache/spark/commit/f21b6936e84c98cea69c2ed7835313fcb01dcf71).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class OrcHadoopWriter(@transient jobConf: JobConf) extends SparkHadoopWriter(jobConf) `

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27470/
Test FAILed.

SparkQA commented:
[Test build #27469 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27469/consoleFull) for   PR 3753 at commit [`9d7c082`](https://github.com/apache/spark/commit/9d7c0826a5d021015b4382559919ac1e1d582d31).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class OrcHadoopWriter(@transient jobConf: JobConf) extends SparkHadoopWriter(jobConf) `

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27469/
Test PASSed.

scwf commented:
@liancheng and @marmbrus , i removed the partitioned support for orc tables and added write interface based on the newly introduced write api, can you help review this? thanks

krise3k commented:
+1

scwf commented:
Retest this please

SparkQA commented:
[Test build #30621 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30621/consoleFull) for   PR 3753 at commit [`f21b693`](https://github.com/apache/spark/commit/f21b6936e84c98cea69c2ed7835313fcb01dcf71).

SparkQA commented:
[Test build #30621 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30621/consoleFull) for   PR 3753 at commit [`f21b693`](https://github.com/apache/spark/commit/f21b6936e84c98cea69c2ed7835313fcb01dcf71).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class OrcHadoopWriter(@transient jobConf: JobConf) extends SparkHadoopWriter(jobConf) `

 * This patch does not change any dependencies.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30621/
Test FAILed.

scwf committed:
fix style

SparkQA commented:
[Test build #30634 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30634/consoleFull) for   PR 3753 at commit [`956c095`](https://github.com/apache/spark/commit/956c095a7de9a9d02e287005d36b6321c05ad951).

SparkQA commented:
[Test build #30634 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30634/consoleFull) for   PR 3753 at commit [`956c095`](https://github.com/apache/spark/commit/956c095a7de9a9d02e287005d36b6321c05ad951).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class OrcHadoopWriter(@transient jobConf: JobConf) extends SparkHadoopWriter(jobConf) `

 * This patch does not change any dependencies.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30634/
Test FAILed.

scwf committed:
Merge branch 'master' of https://github.com/apache/spark into orc-datasourceapi

scwf committed:
fix compile error

SparkQA commented:
[Test build #30645 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30645/consoleFull) for   PR 3753 at commit [`0dd36ee`](https://github.com/apache/spark/commit/0dd36ee9d3e637a24ec0b07f5388eabcc0f3a6f4).

SparkQA commented:
[Test build #30645 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30645/consoleFull) for   PR 3753 at commit [`0dd36ee`](https://github.com/apache/spark/commit/0dd36ee9d3e637a24ec0b07f5388eabcc0f3a6f4).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class OrcHadoopWriter(@transient jobConf: JobConf) extends SparkHadoopWriter(jobConf) `

 * This patch does not change any dependencies.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30645/
Test FAILed.

SparkQA commented:
[Test build #30648 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30648/consoleFull) for   PR 3753 at commit [`9788b85`](https://github.com/apache/spark/commit/9788b85a6e9824249851c5962fc5b9ad80521962).

SparkQA commented:
[Test build #30648 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30648/consoleFull) for   PR 3753 at commit [`9788b85`](https://github.com/apache/spark/commit/9788b85a6e9824249851c5962fc5b9ad80521962).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch **removes the following dependencies:**
   * `RoaringBitmap-0.4.5.jar`
   * `activation-1.1.jar`
   * `akka-actor_2.10-2.3.4-spark.jar`
   * `akka-remote_2.10-2.3.4-spark.jar`
   * `akka-slf4j_2.10-2.3.4-spark.jar`
   * `aopalliance-1.0.jar`
   * `arpack_combined_all-0.1.jar`
   * `avro-1.7.7.jar`
   * `breeze-macros_2.10-0.11.2.jar`
   * `breeze_2.10-0.11.2.jar`
   * `chill-java-0.5.0.jar`
   * `chill_2.10-0.5.0.jar`
   * `commons-beanutils-1.7.0.jar`
   * `commons-beanutils-core-1.8.0.jar`
   * `commons-cli-1.2.jar`
   * `commons-codec-1.10.jar`
   * `commons-collections-3.2.1.jar`
   * `commons-compress-1.4.1.jar`
   * `commons-configuration-1.6.jar`
   * `commons-digester-1.8.jar`
   * `commons-httpclient-3.1.jar`
   * `commons-io-2.1.jar`
   * `commons-lang-2.5.jar`
   * `commons-lang3-3.3.2.jar`
   * `commons-math-2.1.jar`
   * `commons-math3-3.4.1.jar`
   * `commons-net-2.2.jar`
   * `compress-lzf-1.0.0.jar`
   * `config-1.2.1.jar`
   * `core-1.1.2.jar`
   * `curator-client-2.4.0.jar`
   * `curator-framework-2.4.0.jar`
   * `curator-recipes-2.4.0.jar`
   * `gmbal-api-only-3.0.0-b023.jar`
   * `grizzly-framework-2.1.2.jar`
   * `grizzly-http-2.1.2.jar`
   * `grizzly-http-server-2.1.2.jar`
   * `grizzly-http-servlet-2.1.2.jar`
   * `grizzly-rcm-2.1.2.jar`
   * `groovy-all-2.3.7.jar`
   * `guava-14.0.1.jar`
   * `guice-3.0.jar`
   * `hadoop-annotations-2.2.0.jar`
   * `hadoop-auth-2.2.0.jar`
   * `hadoop-client-2.2.0.jar`
   * `hadoop-common-2.2.0.jar`
   * `hadoop-hdfs-2.2.0.jar`
   * `hadoop-mapreduce-client-app-2.2.0.jar`
   * `hadoop-mapreduce-client-common-2.2.0.jar`
   * `hadoop-mapreduce-client-core-2.2.0.jar`
   * `hadoop-mapreduce-client-jobclient-2.2.0.jar`
   * `hadoop-mapreduce-client-shuffle-2.2.0.jar`
   * `hadoop-yarn-api-2.2.0.jar`
   * `hadoop-yarn-client-2.2.0.jar`
   * `hadoop-yarn-common-2.2.0.jar`
   * `hadoop-yarn-server-common-2.2.0.jar`
   * `ivy-2.4.0.jar`
   * `jackson-annotations-2.4.0.jar`
   * `jackson-core-2.4.4.jar`
   * `jackson-core-asl-1.8.8.jar`
   * `jackson-databind-2.4.4.jar`
   * `jackson-jaxrs-1.8.8.jar`
   * `jackson-mapper-asl-1.8.8.jar`
   * `jackson-module-scala_2.10-2.4.4.jar`
   * `jackson-xc-1.8.8.jar`
   * `jansi-1.4.jar`
   * `javax.inject-1.jar`
   * `javax.servlet-3.0.0.v201112011016.jar`
   * `javax.servlet-3.1.jar`
   * `javax.servlet-api-3.0.1.jar`
   * `jaxb-api-2.2.2.jar`
   * `jaxb-impl-2.2.3-1.jar`
   * `jcl-over-slf4j-1.7.10.jar`
   * `jersey-client-1.9.jar`
   * `jersey-core-1.9.jar`
   * `jersey-grizzly2-1.9.jar`
   * `jersey-guice-1.9.jar`
   * `jersey-json-1.9.jar`
   * `jersey-server-1.9.jar`
   * `jersey-test-framework-core-1.9.jar`
   * `jersey-test-framework-grizzly2-1.9.jar`
   * `jets3t-0.7.1.jar`
   * `jettison-1.1.jar`
   * `jetty-util-6.1.26.jar`
   * `jline-0.9.94.jar`
   * `jline-2.10.4.jar`
   * `jodd-core-3.6.3.jar`
   * `json4s-ast_2.10-3.2.10.jar`
   * `json4s-core_2.10-3.2.10.jar`
   * `json4s-jackson_2.10-3.2.10.jar`
   * `jsr305-1.3.9.jar`
   * `jtransforms-2.4.0.jar`
   * `jul-to-slf4j-1.7.10.jar`
   * `kryo-2.21.jar`
   * `log4j-1.2.17.jar`
   * `lz4-1.2.0.jar`
   * `management-api-3.0.0-b012.jar`
   * `mesos-0.21.0-shaded-protobuf.jar`
   * `metrics-core-3.1.0.jar`
   * `metrics-graphite-3.1.0.jar`
   * `metrics-json-3.1.0.jar`
   * `metrics-jvm-3.1.0.jar`
   * `minlog-1.2.jar`
   * `netty-3.8.0.Final.jar`
   * `netty-all-4.0.23.Final.jar`
   * `objenesis-1.2.jar`
   * `opencsv-2.3.jar`
   * `oro-2.0.8.jar`
   * `paranamer-2.6.jar`
   * `parquet-column-1.6.0rc3.jar`
   * `parquet-common-1.6.0rc3.jar`
   * `parquet-encoding-1.6.0rc3.jar`
   * `parquet-format-2.2.0-rc1.jar`
   * `parquet-generator-1.6.0rc3.jar`
   * `parquet-hadoop-1.6.0rc3.jar`
   * `parquet-jackson-1.6.0rc3.jar`
   * `protobuf-java-2.4.1.jar`
   * `protobuf-java-2.5.0-spark.jar`
   * `py4j-0.8.2.1.jar`
   * `pyrolite-2.0.1.jar`
   * `quasiquotes_2.10-2.0.1.jar`
   * `reflectasm-1.07-shaded.jar`
   * `scala-compiler-2.10.4.jar`
   * `scala-library-2.10.4.jar`
   * `scala-reflect-2.10.4.jar`
   * `scalap-2.10.4.jar`
   * `scalatest_2.10-2.2.1.jar`
   * `slf4j-api-1.7.10.jar`
   * `slf4j-log4j12-1.7.10.jar`
   * `snappy-java-1.1.1.7.jar`
   * `spark-bagel_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-catalyst_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-core_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-graphx_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-launcher_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-mllib_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-network-common_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-network-shuffle_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-repl_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-sql_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-streaming_2.10-1.4.0-SNAPSHOT.jar`
   * `spire-macros_2.10-0.7.4.jar`
   * `spire_2.10-0.7.4.jar`
   * `stax-api-1.0.1.jar`
   * `stream-2.7.0.jar`
   * `tachyon-0.5.0.jar`
   * `tachyon-client-0.5.0.jar`
   * `uncommons-maths-1.2.2a.jar`
   * `unused-1.0.0.jar`
   * `xmlenc-0.52.jar`
   * `xz-1.0.jar`
   * `zookeeper-3.4.5.jar`

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30648/
Test FAILed.

scwf commented:
Retest this please.

SparkQA commented:
[Test build #30657 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30657/consoleFull) for   PR 3753 at commit [`9788b85`](https://github.com/apache/spark/commit/9788b85a6e9824249851c5962fc5b9ad80521962).

SparkQA commented:
[Test build #30657 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30657/consoleFull) for   PR 3753 at commit [`9788b85`](https://github.com/apache/spark/commit/9788b85a6e9824249851c5962fc5b9ad80521962).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30657/
Test PASSed.

scwf commented on core/src/main/scala/org/apache/spark/SparkHadoopWriter.scala:
```diff
@@ -41,20 +41,20 @@ class SparkHadoopWriter(@transient jobConf: JobConf)
+  @transient protected var format: OutputFormat[AnyRef,AnyRef] = null
+  @transient protected var committer: OutputCommitter = null
+  @transient protected var jobContext: JobContext = null
+  @transient protected var taskContext: TaskAttemptContext = null
```
I changed the scope of var/def of SparkHadoopWriter to reuse these code in orc writting api implementation

scwf commented:
To make ORC datasource clean and easy to review, i will split it to three part of work, each one should be a PR.
1 orc datasource api support including read/write implementation, no partitioned support
2 filter push down optimization
3 partitioning support

This is the PR for the first point.

/cc @marmbrus @liancheng

scwf commented:
ping

scwf commented:
i am closing this in favor of #6914
